### ISSUE TITLE
Run E2E tests as part of the Jenkins PR build

### ISSUE THREAD

**Issue by lukaszfryc** [status-im/status-react#3740]
### Description

Currently, E2E tests for PRs are run in a separate Jenkins build. Instead, they should be run as a last step of the PR build using the .apk that was created in the previous step. 

<img src="https://user-images.githubusercontent.com/7532782/38188630-105d12e4-365d-11e8-8ede-3db3da9e4ae2.png" width="400">

This way we could:
- Be sure what build (commit id) was used
- Have everything into one place and link github test result to the jenkins build url

For now, the test results can be ignored to not stop the build when there is a failure. In the near future though, the test failure should block the PR from merging (cc @yevh-berdnyk)

### Solution

- [ ] Add running tests as another step into PR pipeline. Ignore the results for now 
- [ ] Use .apk from previous step

**Comment by lukaszfryc** [comment]
@antdanchenko @yevh-berdnyk wdyt? Having everything in one place (one pipeline) seems like a good idea. Especially if we want tests to block the PRs in Q2.

**Comment by antdanchenko** [comment]
@lukaszfryc It seems that you know my opinion since we already discussed that

Anyway, instead of running in a PR job, would be better to run in our own job where we can build the app from a PR branch
Steps for building the app for tests are different, so we won't be able to reuse the app from the PR job in nearest future. Also running in the PR job will add ~20 minutes of execution time

_The only issue is additional mac os machine for e2e tests instead of linux1 node. The machine is required for building the app in `e2e tests` job_

**Comment by yevh-berdnyk** [comment]
@lukaszfryc I think the only step which could be added to PR jobs is the 'first layer' of tests which will block the PR. But anyway we need to rebuild the app for other tests (with Ropsten) so it would be better to run them separately and don't duplicate even few tests in PR job

**Comment by lukaszfryc** [comment]
> Anyway, instead of running in a PR job, would be better to run in our own job where we can build the app from a PR branch
Steps for building the app for tests are different, so we won't be able to reuse the app from the PR job in nearest future. Also running in the PR job will add ~20 minutes of execution time

Yes, I see that we use `.env.e2e` environments to build it. So, you are right. We can't reuse the .apk from "Deploy (Android)" stage.

But, I think it would be better idea to run everything that is needed for a PR in single pipeline. It does not matter that the execution time will be longer since we require the tests to execute before merging the PR. Of course, only the "first layer" tests will block the PR. The rest of tests will only provide feedback. 

Having single pipeline will simplify things. We will have everything that's required for PR in one place. It seems really natural.

**Comment by lukaszfryc** [comment]
Here is how I see the pull request pipeline:

|Step|Current pipeline|New pipeline|
|-|-|-|
|1|Git & Dependencies|Git & Dependencies|
|2|Tests|Tests|
|3|Build|Build|
|4|Build (Android)|Build (Android)|
|5|Deploy (Android)|Deploy (Android)|
|6|Build (iOS)|Build (iOS)|
|7|Deploy (iOS)|Deploy (iOS)|
|8|Slack Notification|Slack Notification|
|9|Build (Android) for e2e testing|Build (Android) for e2e testing|
|10|Upload apk for e2e tests|Upload apk for e2e tests|
|11||Smoke e2e tests (Android)|
|12||Regression e2e tests (Android)|

So the only changes are 11th (first layer tests that are stable) and 12th (selected tests from the regression suite) steps. In both of these steps we would use url to apk from 9th. The 11th would block the PR if failed. 12th would not. It would only provide test results in the PR.

Of course, adding the 2 steps means longer build time. But, both of these steps need to happen before PR can be merged. So, it makes sense for to keep them in the PR pipeline. 
Also, no one needs to wait for the build to finish before starting testing the app since 8th step sends slack notifications with the builds.

**Comment by lukaszfryc** [comment]
One reason to not have it in a single pipeline is that the PR Jenkins Job can't execute few builds in parallel. Meaning, the PRs will be blocked by long-running builds. But, if that's the case we should just increase resources in this Job.

**Comment by lukaszfryc** [comment]
Here is how we could extend `Jenkinsfile`

```
    stage('e2e smoke tests (Android)') {
      node('linux1') {
        // sauce('12e007ad-48cf-4c20-92f3-b923bb5641bd') {
        // ...
      }
    }

    try {
      stage('e2e regression tests (Android)') {
        node('linux1') {
          // ...
        }
      }
    } catch (e) {
      // do not block PR when e2e regression tests failed
    }
```

So, the `macos1` node would execute `linux1` node to run the appium tests. But, `macos1` would need to wait until `linux1` is done. In order to simplify it, I think we could execute appium tests on `macos1` node and remove `linux1` completely. `macos1` image would contain all necessary dependencies. 

I'd also add couple of more `macos1` slaves to enable more parallel Jenkins builds.

**Comment by lukaszfryc** [comment]
@antdanchenko @yevh-berdnyk any thoughts? I think at least the smoke stage (first level tests) maybe necessary in the PR build since they will block it if failed.

**Comment by antdanchenko** [comment]
you can try in order to see how it goes, but please define the suite which should block a PR first

and for me it makes sense to execute `Smoke e2e tests (Android)` before `Slack Notification`, WDYT?

please note, that beside increasing an amount of slaves we should take care about number of sessions (on SauceLab) for tests to be executed (right now 24 sessions can be used in the same time and a lot of tests are using 2+ sessions per test)

**Comment by antdanchenko** [comment]
Tests are going to run and spam results to a PR even if it's not in `TO TEST` in your variation of the pipeline

We tried that before and it was too noisy, so we are running them only if a PR is in `TO TEST`. So I would recommend not to run them if a PR is not in `TO TEST`

**Comment by lukaszfryc** [comment]
> but please define the suite which should block a PR first

@yevh-berdnyk could you take it?

> and for me it makes sense to execute Smoke e2e tests (Android) before Slack Notification, WDYT?
> So I would recommend not to run them if a PR is not in TO TEST

It makes sense. It we say that critial things do not work (smoke test), there is no need to announce the builds. But, it means that smoke tests should run with every commit. To be honest, I'd start with running them only when PR is in TO TEST and leave the slack notification stage as it is. Then if we see that smoke tests give value, are stable, fast, etc. we can make run them with every commit and only then send slack notification.

> please note, that beside increasing an amount of slaves we should take care about number of sessions (on SauceLab) for tests to be executed (right now 24 sessions can be used in the same time and a lot of tests are using 2+ sessions per test)

Yes!

----

Summing up - it does not look like an easy task. But, if we want to have smoke tests blocking PRs in Q2 we should do it. But, it's not top priority now. @yevh-berdnyk I'd start with defining the list of scenarios for the smoke test. Then, let's make sure they are stable (like in 100 PRs they were stable in 95%). Only then we will go back to this issue and will start to block the PRs.
